src/main/java/com/tananaev/adblib/AdbConnection.java

```diff
@@ -285,11 +285,18 @@ public boolean connect(long timeout, TimeUnit unit) throws IOException, Interrup
 
         /* Wait for the connection to go live */
         synchronized (this) {
-            if (!connected)
-                wait(unit.toMillis(timeout));
+            long timeoutDurationMillis = unit.toMillis(timeout);
+            boolean hasTimeout = timeoutDurationMillis != 0;
+            long timeoutEndMillis = System.currentTimeMillis() + timeoutDurationMillis;
+            long remainingTimeoutMillis = timeoutDurationMillis;
+
+            while (!connected && connectAttempted && (!hasTimeout || remainingTimeoutMillis > 0)) {
```

Comment:
Do we actually need to check both `!hasTimeout || remainingTimeoutMillis > 0`? Wouldn't `remainingTimeoutMillis > 0` be enough?

Reply:
Without the `hasTimeout` check, then if `timeout` is `0` (meaning infinite timeout) then `remainingTimeoutMillis` will be `0` and the loop will never be entered, so the connection will fail immediately

Reply:
The code would be heaps easier to read if AdbLib used something from `java.util.concurrent`, which takes care of timeouts and spurious wake-ups for us, instead of `wait` and `notify`, but I feel like it'd be a lot of work to make the change

Reply:
Maybe we can use `Long.MAX...` as the default instead of 0 to indicate "no timeout"?

Reply:
Good idea, that makes the code nice and clean. I've made the change in c1be70f.